### ISSUE TITLE
with-rss2 macro broken

### ISSUE THREAD

**Issue by VitoVan** [VitoVan/xml-emitter#6]
### How to reproduce:

```lisp
(with-output-to-string (s) (with-rss2 (s)))
```

It will fail as:

```lisp
; in: WITH-OUTPUT-TO-STRING (S)
;     (WITH-RSS2 (S))
; --> WITH-XML-OUTPUT LET WITH-TAG LET PROGN XML-EMITTER::START-TAG 
; ==>
;   (("version" "2.0"))
; 
; caught ERROR:
;   illegal function call
; 
; compilation unit finished
;   caught 1 ERROR condition
; Evaluation aborted on #<SB-INT:COMPILED-PROGRAM-ERROR {100926A0D3}>.
```

### Try digging a little deeper:

```lisp
(macroexpand '(with-rss2 (s)))
```

this macro expanded to:

```lisp
(LET ((XML-EMITTER::*XML-OUTPUT-STREAM* S))
  (FORMAT XML-EMITTER::*XML-OUTPUT-STREAM*
          "<?xml version=\"1.0\" encoding=~S~:[~; standalone=\"~:[no~;yes~]\"~]?>~%"
          "ISO-8859-1" NIL NIL)
  (WITH-TAG ("rss" (("version" "2.0")))
    (WITH-TAG ("channel"))))
```

### When do we changed this?

https://github.com/VitoVan/xml-emitter/pull/5/files#diff-5fe6d35e26e9626492ab0952c60beb81R68

It was written as:

```lisp
(with-tag ("rss" '(("version" "2.0"))) ...)
```

now it is expanded as:

```lisp
(WITH-TAG ("rss" (("version" "2.0")))
    (WITH-TAG ("channel")))
```

see the `'(("version" "2.0"))` vs `(("version" "2.0"))`

### How to fix?

1. I tried to add double quotes and it works:

    ```lisp
    (defmacro with-rss2 ((stream &key (encoding "ISO-8859-1") (attrs '(("version" "2.0")))) &body body)
    ```

    to

    ```lisp
    (defmacro with-rss2 ((stream &key (encoding "ISO-8859-1") (attrs ''(("version" "2.0")))) &body body)
    ```

2. Or add quote inside the macro

    ```lisp
    (with-tag ("rss" ,attrs)
           (with-tag ("channel")
             ,@body))
    ``` 

    to

    ```lisp
    (with-tag ("rss" ',attrs)
           (with-tag ("channel")
             ,@body))
    ```

### Question

Should we fix it with extra quotes? Is that ugly? Or am I wrong about this?

Related: https://github.com/VitoVan/xml-emitter/pull/5

And I think it's time to add tests

**Comment by VitoVan** [comment]
@iamFIREcracker Hi Matteo Landi, feel free to check on this, if you have time and interests. Your thoughts would be appreciated very much. 

And happy new year to you.

**Comment by VitoVan** [comment]
Here's the PR I tried: https://github.com/VitoVan/xml-emitter/pull/8

**Comment by iamFIREcracker** [comment]
I was trying to understand why I did not catch this earlier, but then realized I was always overriding `ATTRS`: https://github.com/iamFIREcracker/plan-rss/blob/master/src/main.lisp#L230-L231

Anyway, will have another look at the MR either today or tomorrow and get back to you with some feedback (still not sure about the _best_ fix though).

**Comment by VitoVan** [comment]
@iamFIREcracker 

Thanks Matteo Landi, take your time.

This bug has been here for a while, no one complains before, so I think we can take it easy 😸

**Comment by iamFIREcracker** [comment]
@VitoVan: what compiler are you using?  I am testing with SBCL 2.0.0 but it's not spitting any WARNING/ERROR -- even though the call to `MACROEXPAND-1` generates the same output as yours (i.e. without `QUOTE`, which looks wrong).

PS. Apologies for the late reply but was off with very limited internet connectivity, and I managed to look into this only today.

**Comment by VitoVan** [comment]
@iamFIREcracker I'm using SBCL 1.5.9, and I will check this again on SBCL 2.0.0.

PS. Don't worry about late reply, life always changing without our authorization. BTW, it's glad to hear you.

**Comment by VitoVan** [comment]
I just tried this on SBCL 2.0.0, with no luck:

```lisp
XML-EMITTER> (lisp-implementation-version)
"2.0.0"
XML-EMITTER> (with-output-to-string (s) (with-rss2 (s)))
; in: WITH-OUTPUT-TO-STRING (S)
;     (XML-EMITTER:WITH-RSS2 (XML-EMITTER::S))
; --> XML-EMITTER:WITH-XML-OUTPUT LET XML-EMITTER:WITH-TAG LET PROGN 
; --> XML-EMITTER::START-TAG 
; ==>
;   (("version" "2.0"))
; 
; caught ERROR:
;   illegal function call
; 
; compilation unit finished
;   caught 1 ERROR condition
; Evaluation aborted on #<SB-INT:COMPILED-PROGRAM-ERROR {10037C9213}>.
XML-EMITTER> 
```

**Comment by iamFIREcracker** [comment]
All right, I was being dumb -- I forgot to run `(ql:update-dist :xml-emitter)` so it was still using the last snapshot from 2018.

Anyway, I will shortly come up with an PR to:

- Fix the problem (I think I would go with double quoting the `&optional` argument)
- Add a few tests (I have some experience with `1AM`: http://quickdocs.org/1am/)
- Add CI support (I used TravisCI in the past: https://github.com/iamFIREcracker/cg/blob/master/.travis.yml)

Let me know your thoughts about this.

**Comment by iamFIREcracker** [comment]
@VitoVan: https://github.com/VitoVan/xml-emitter/pull/9 -- thoughts?  It needs some refactoring, and support for Travis CI, that at least should give you an idea of where I am heading

**Comment by VitoVan** [comment]
@iamFIREcracker 

My thoughts: You're heading to a wonderful world with perfect xml-emitter loaded while I'm sleeping.

**Comment by VitoVan** [comment]
Travis CI is configured here:

https://travis-ci.com/VitoVan/xml-emitter

Let me know if there's anything need to be done

**Comment by VitoVan** [comment]
And for Travis CI, I used to compile the latest SBCL from source like this:

https://github.com/VitoVan/cl-pkr/blob/master/.travis.yml

since SBCL do not always ship the latest binary as we ask.

PS. You can leave this shit work to me if it's not enjoyable to you.

**Comment by iamFIREcracker** [comment]
Did some progress:

- switched to plain ros for CI, instead of cl-travis (uses CIM, now deprecated)
- Enhanced Makefile to gracefully handle failures (the various lisp implementations would start in interactive mode, so in case of errors the debugger would be invoked, causing the job to hang)

Here is an example of how a failed build would look like:

https://travis-ci.org/iamFIREcracker/xml-emitter/builds/637689811

While here of a successful one:

https://travis-ci.org/iamFIREcracker/xml-emitter/builds/637695171

Tomorrow I will do some small refactoring on the tests file (to remove that copy-pasta), but other than that, I think we should be good to go!

**Comment by VitoVan** [comment]
@iamFIREcracker 

You have done a great job, Matteo Landi.

I always hate roswell, but you make it feels right.

The hate of roswell came from a very wired mind loop, it has my envy and ego involved (and of course some far away but related technical thoughts), I should get over it.

**Comment by iamFIREcracker** [comment]
When I approached Common Lisp for the first time, I did it with roswell, and honestly did not run into many issues with it; then I realized I was almost always using SBCL all the time, so decided to ditch roswell, and switch to pure Common Lisp implementations instead.  Now, Common Lisp programming is just a hobby for me, so I am well aware that my view could be limited about it.

Still, roswell does an amazing job with making CI testing easy, especially if interested in testing with multiple implementation, and because of it (and the lack of experience with other implementation managers), I suggest we go ahead with this approach.
